### ISSUE TITLE
Dependencies must be provided in the topological order in order to avoid needless recompilation

### ISSUE THREAD

**Issue by maximedenes** [coq/coq-bench#3]
The current script does something silly: if there are `n` packages, it can perform `n(n+1) / 2` builds because of dependencies.

Why do we remove packages after we benched them? If they stayed installed, they could be used as dependencies for the next packages without reinstalling them. We could then pass to the script the names of packages in topological order.

@matejkosik can you take care of this? The EConstr bench failed because of a CI issue. It would help if we cut down the time of the bench.

**Comment by maximedenes** [comment]
Or maybe it is just that we didn't put packages in the right order in the previous examples we run.

**Comment by maximedenes** [comment]
I relaunched a build for EConstr with mathcomp dependencies sorted in topological order. Please don't interfere with the build. I'll report here to confirm the problem or not.

**Comment by ghost** [comment]
I understand what you mean. The silly behavior represents the simplest way how to satisfy the original requirement that we did not want to force the user to provide OPAM packages in in a topological order.

Maybe we can really (for now) relax that (useful) requirement and assume that the OPAM packages are provided in the topological order. This would immediatelly cut down the benchmarking time 10x (maybe).

We can address the original requirement (ability to provide OPAM packages in arbitrary order) later (or not at all, if it will not be relevant anymore).

**Comment by ghost** [comment]
To satisfy both requirements:
- not to remove benchmarked OPAM packages that are required by other OPAM packages we want to benchmark later
- enable the user to provide OPAM packages in arbitrary order

would require different approach. I do not think I can use `opam` for that.

**Comment by maximedenes** [comment]
Yes, but if packages are provided in the topological order, even the script of today is much faster, right? So why isn't that documented? And why where the first example builds run with packages in the wrong order??

Also, why don't we use `opam install --show-actions` to do the topological sort for us?

**Comment by maximedenes** [comment]
Something puzzles me. Is it right that the script removes packages *after* benchmarking them? Why do we want that? I would have expected that we remove *before* benchmarking in case they were added as a dependency by a previous package.

**Comment by ghost** [comment]
@maximedenes Now, when you mention it, that is what I should be doing.

**Comment by maximedenes** [comment]
But there is still something I don't get in the current script: if a package is installed as a dependency, what removes it before benchmarking it?

**Comment by ghost** [comment]
[this](https://github.com/coq/coq-bench/blob/master/two_points_on_the_same_branch.sh#L462-L463) and [this](https://github.com/coq/coq-bench/blob/master/two_points_on_the_same_branch.sh#L473-L474).

Those two OPAM-roots are created before benchmarking. Each of those OPAM-roots contains the right version of Coq (for the one and the other commit we were interested in).

**Comment by maximedenes** [comment]
Oh, you delete the opam root each time we measure a package??

**Comment by ghost** [comment]
At the moment, yes.

(although if I have 10 iterations, I reuse a given OPAM-root 10-times)

But after benchmarking of a given OPAM-package is finished, I throw away the OPAM root.

**Comment by maximedenes** [comment]
That sounds like a very bad design. Can you fix it?

1. remove the packages *before* benchmarking them, not *after*
2. do not delete the OPAM-root when the benchmarking of a package is finished
3. use `opam install --show-actions` to perform the topological sort

If you can get 1 and 2 quickly and push it, I'll rerun the EConstr bench. 3 can come later.

**Comment by ghost** [comment]
Ok.

**Comment by ghost** [comment]
I've pushed the corresponding changes (look at the [last commit](https://github.com/coq/coq-bench/commit/d075ad0e2a434ef0a203a4f0de8054622cc22cf5); the previous commits are related to something else I testing right now).

The last commit is completely untested. I am not sure if it does what we want.

**Comment by maximedenes** [comment]
Thanks, but there is still something I don't understand. This commit comments that we assume the user provides packages in topological order, and hence it is a hack. Why?! What would go wrong if they are not in the topological order?

**Comment by ghost** [comment]
At the moment (i.e. after [this](https://github.com/coq/coq-bench/commit/d075ad0e2a434ef0a203a4f0de8054622cc22cf5) commit), if we do not provide OPAM packages in a topological order, the benchmarking will take longer.

Imagine the script is about to benchmark a package `A`.

In theory, it could already have been installed, so I must do (before benchmarking) `opam uninstall A`.

Now, if something else was already installed and it depends on `A`, it will be removed.

That is a shame because that other thing might be needed as a dependency of something which we want to benchmark later. In this case, it will be removed just to be later reinstalled.

In short, it would prolong the benchmarking.

**Comment by ghost** [comment]
I think I must refine the corresponding comment. Although the plan is to remove that limitation anyway ...
(as you noted, `opam install --show-actions` might be the only thing that we need)

I need to investigate that.

**Comment by maximedenes** [comment]
Yes, it will take longer, but it will still be running fine, so that is not an assumption. And it is also not a hack, it always behaves better or identically to the previous solution.

Now, it would be good to try to do the sorting by calling `opam install --show-actions`. What made you reject that solution? Was there a catch?

**Comment by ghost** [comment]
I remember that I have found this option, I looked at it, and in the end I saw that I couldn't use it.

(this is true, I think, in a different context --- when I was implementing `opam-install-*` jobs on Jenkins. In my mind it was rejected even though in this context it may be usable.)

In any case, I was planning to return to this problem. But many (much more simple) things did not work, which was frustrating.

Also, the fact that the job, as a whole, was failing whenever a selected OPAM package was not `opam install`-able, I found more emberassing. When the job runs for 20 hours and then says, good bye (this tiny thing is not installble).

**Comment by maximedenes** [comment]
I would say indeed, that resilience to errors is now the priority. It should be reasonably easy to do, right?

Computing the topological sort using OPAM would be next.

**Comment by ghost** [comment]
@maximedenes 
> I would say indeed, that resilience to errors is now the priority. It should be reasonably easy to do, right?

I think so. The `opam_install.sh` must go and the loop (if we need any) must be outside the `time` and `perf` commands.

@maximedenes 
> Computing the topological sort using OPAM would be next.
Ok. Btw. the default OPAM packages are topologically sorted, anyway (unless I made a mistake when I was checking this).

**Comment by ghost** [comment]
There is a bug in the [last commit](https://github.com/coq/coq-bench/commit/d075ad0e2a434ef0a203a4f0de8054622cc22cf5).

For the moment, don't use `number_of_iterations > 1`.

**Comment by ghost** [comment]
This issue is not yet resolved (we will have to take advantage of `opam install --show-actions`) but it is at least mitigated.

Unless you specify the OPAM packages for benchmarking in non-topological order, no OPAM package is needlessly recompiled.

**Comment by JasonGross** [comment]
If you don't care about preserving the relative order of independent dependencies, you should be able to implement this by replacing
```bash
coq_opam_packages_on_separate_lines=$(echo "$coq_opam_packages" | sed 's/ /\n/g')
```
with
```bash
coq_opam_packages_on_separate_lines=$(opam install --show-actions $coq_opam_packages | grep -o '∗\s\+install\s\+[^ ]*' | sed 's/^∗\s\+install\s*//g')
```
